lte/cloud/go/services/subscriberdb/storage/digest.go

```diff
@@ -0,0 +1,222 @@
+ limitations under the License.
+*/
+
+package storage
+
+import (
+	"database/sql"
+
+	"magma/orc8r/cloud/go/clock"
+	"magma/orc8r/cloud/go/sqorc"
+
+	"github.com/Masterminds/squirrel"
+	"github.com/pkg/errors"
+)
+
+type digestLookup struct {
+	db      *sql.DB
+	builder sqorc.StatementBuilder
+}
+
+type digestInfo struct {
+	network         string
+	digest          string
+	lastUpdatedTime int64
+}
+
+type DigestInfoSet []digestInfo
+
+// GetNetworks returns a list of network IDs for all digests in DigestInfoSet.
+func (digestInfoSet DigestInfoSet) GetNetworks() []string {
+	ret := []string{}
+	for _, digestInfo := range digestInfoSet {
+		ret = append(ret, digestInfo.network)
+	}
+	return ret
+}
+
+type DigestLookup interface {
+	// Initialize the backing store.
+	Initialize() error
+
+	// GetDigests returns a list of digests of that satisfy the filtering criteria
+	// specified by the arguments.
+	// Caveats:
+	// 1. If networks is empty, returns digests for all networks.
+	// 2. earliestUpdateTime is recorded in unix seconds. If a positive integer,
+	// filter for all digests that were last updated earlier than this time
+	// (i.e. outdated digests); otherwise, the time filter is not applied.
+	GetDigests(networks []string, earliestUpdateTime int64) (DigestInfoSet, error)
+
+	// SetDigest creates/updates the subscribers digest for a particular network.
+	SetDigest(network string, digest string) error
+
+	// DeleteDigests removes digest rows by network IDs.
+	DeleteDigests(networks []string) error
+}
+
+const (
+	digestLookupTableName = "subscriberdb_flat_digests"
+
+	digestLookupNidCol             = "network_id"
+	digestLookupDigestCol          = "digest"
+	digestLookupLastUpdatedTimeCol = "last_updated_at"
+)
+
+func NewDigestLookup(db *sql.DB, builder sqorc.StatementBuilder) DigestLookup {
+	return &digestLookup{db: db, builder: builder}
+}
+
+func (l *digestLookup) Initialize() error {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		_, err := l.builder.CreateTable(digestLookupTableName).
+			IfNotExists().
+			Column(digestLookupNidCol).Type(sqorc.ColumnTypeText).NotNull().EndColumn().
+			Column(digestLookupDigestCol).Type(sqorc.ColumnTypeText).NotNull().EndColumn().
+			Column(digestLookupLastUpdatedTimeCol).Type(sqorc.ColumnTypeBigInt).NotNull().EndColumn().
+			PrimaryKey(digestLookupNidCol).
+			RunWith(tx).
+			Exec()
+		return nil, errors.Wrap(err, "initialize digest lookup table")
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+func (l *digestLookup) GetDigests(networks []string, earliestUpdateTime int64) (DigestInfoSet, error) {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		filters := squirrel.And{}
+		if len(networks) > 0 {
+			filters = append(filters, squirrel.Eq{digestLookupNidCol: networks})
+		}
+		if earliestUpdateTime > 0 {
+			filters = append(filters, squirrel.Lt{digestLookupLastUpdatedTimeCol: earliestUpdateTime})
+		}
```

Comment:
From above: apply this filter unconditionally

Reply:
Since we can use `clock.now` for the "lastUpdatedBefore" filter value, we can apply this filter unconditionally

Reply:
Gotcha, seems like using `clock.now` for returning all digests is more intuitive than using `0` because it makes the special caveat unnecessary